explorer/interpreter/type_checker.h

```diff
@@ -565,6 +565,10 @@ class TypeChecker {
   auto InterpExp(Nonnull<const Expression*> e)
       -> ErrorOr<Nonnull<const Value*>>;
 
+  // Deduces concrete type for 'type' based on 'expected'
+  auto Deduce(Nonnull<const Value*> type, Nonnull<const Value*> expected)
```

Comment:
The name and comment make this function sound far more general than it seems like the implementation actually supports. It also makes it sound like this is redundant with `ArgumentDeduction`. Is it possible to use `ArgumentDeduction` for this purpose? If not, can the name/comment be changed to make it clearer how they're different?